### PR TITLE
Use Heart Beat API to consume EA client results.

### PR DESCRIPTION
**Pull request by mitogh**
Use the Heart Beath API to pull the status on the progress bar of EA, in this system the poll is done and handling via WordPress instead.

Artifact: http://p.tri.be/iT4IqG

### CODE REVIEW

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 1]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
 	}
 
 	/**
-	 * Handle queue ajax requests
+     * Action to reply every time a heart beat is executed to send the progress of EA if an EA record is present.
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:114]: :no_entry_sign: Tabs must be used to indent lines; spaces are not allowed

<sub>hash: a7e50cbeb69cd719d001e5786874ac6b</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 2]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
+     *
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:115]: :no_entry_sign: Tabs must be used to indent lines; spaces are not allowed

<sub>hash: 752f7c23bd06800ee1669d9a4d14c007</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 3]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
+	 * @param array<string, mixed> $response The current response object
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:116]: :no_entry_sign: Parameter comment must end with a full stop

<sub>hash: 903af9ffcb37e2aa0f1e47a264e6f689</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 4]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
-		// Nonce check
-		$this->ajax_operations->verify_or_exit( $_POST['check'], $this->get_ajax_nonce_action(), $this->get_unable_to_continue_processing_data() );
+		$this->record_id = absint( $data['ea_record'] );
 
-		// Load the queue
-		/** @var \Tribe__Events__Aggregator__Record__Queue_Interface $queue */
-		$queue = $this->queue ? $this->queue : Tribe__Events__Aggregator__Record__Queue_Processor::build_queue( $this->record_id );
-		// We always need to setup the Current Queue
-		$this->queue_processor->set_current_queue( $queue );
+		if ( $this->record_id === 0 ) {
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:128]: :no_entry_sign: Use Yoda Condition checks, you must.

<sub>hash: 7ccc1abb8d6e0c4cf636dbc6a9f3e0ad</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 5]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
-		$percentage = $current_queue->progress_percentage();
-		$this->ajax_operations->exit_data( $this->get_progress_message_data( $current_queue, $percentage, $done ) );
+		return $response;
+	}
+
+	/**
+	 * Handle queue ajax requests
+	 */
+	public function ajax() {
+	    _deprecated_function( __METHOD__, 'TBD' );
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:147]: :no_entry_sign: Tabs must be used to indent lines; spaces are not allowed

<sub>hash: 70c3bd3e8cee83eea7219fa813a0f6fb</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 6]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
+
+		$this->record_id = (int) $_POST['record'];
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:149]: * :no_entry_sign: Processing form data without nonce verification.
* :no_entry_sign: Detected usage of a possibly undefined superglobal array index: $_POST['record']. Use isset() or empty() to check the index exists before using it

<sub>hash: 7455cc53d049d63a1ed9206c700f5197</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 7]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
+
+		// Nonce check
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:151]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 426e751c4dc78e7dd41edead33ecd14b</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 8]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
+		$this->ajax_operations->verify_or_exit( $_POST['check'], $this->get_ajax_nonce_action(), $this->get_unable_to_continue_processing_data() );
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:152]: * :warning: Detected usage of a non-sanitized input variable: $_POST['check']
* :no_entry_sign: Processing form data without nonce verification.
* :no_entry_sign: Detected usage of a possibly undefined superglobal array index: $_POST['check']. Use isset() or empty() to check the index exists before using it

<sub>hash: 27315f957ba769b1be999435a19032ce</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 9]
```diff
@@ -110,35 +111,48 @@ public function render_update_message() {
...
+		$data = $this->get_queue_progress_data();
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:153]: :warning: Equals sign not aligned with surrounding assignments; expected 6 spaces but found 1 space

<sub>hash: bc39f361851bc533c5a2074079c3eb46</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 10]
```diff
@@ -196,6 +210,59 @@ public function get_unable_to_continue_processing_data() {
 	 * @return mixed|string|void
 	 */
 	public function get_progress_message_data( $queue, $percentage, $done ) {
+		_deprecated_function( __METHOD__, 'TBD' );
+
+		return wp_json_encode( $this->get_progress_data($queue, $percentage, $done) );
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:215]: * :no_entry_sign: Expected 1 spaces after opening parenthesis; 0 found
* :no_entry_sign: Expected 1 spaces before closing parenthesis; 0 found
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis

<sub>hash: cae18937af76e184390c0c441b118e1a</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 11]
```diff
@@ -196,6 +210,59 @@ public function get_unable_to_continue_processing_data() {
...
+	 * @since TBD
+	 *
+	 * @return array<string, mixed> An array with the details of the progress bar.
+	 */
+	private function get_queue_progress_data() {
+		if ( $this->record_id <= 0 ) {
+			return [];
+		}
+
+		// Load the queue
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:230]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: d0e730f247632c6910724a444d7e9ab3</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 12]
```diff
@@ -196,6 +210,59 @@ public function get_unable_to_continue_processing_data() {
...
+		/** @var \Tribe__Events__Aggregator__Record__Queue_Interface $queue */
+		$queue = $this->queue ? $this->queue : Tribe__Events__Aggregator__Record__Queue_Processor::build_queue( $this->record_id );
+		// We always need to setup the Current Queue
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:233]: :no_entry_sign: Inline comments must end in full-stops, exclamation marks, or question marks

<sub>hash: 339293a23633d9aada23a257d2ed6464</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 16 error(s)
:warning: 2 warning(s)

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 13]
```diff
@@ -110,35 +111,53 @@ public function render_update_message() {
...
+	 * Handle queue ajax requests
+	 */
+	public function ajax() {
+		_deprecated_function( __METHOD__, 'TBD' );
+
+		$this->record_id = (int) tribe_get_request_var( 'record' );
+
+		// Nonce check.
+		$this->ajax_operations->verify_or_exit(
+			tribe_get_request_var('check'),
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:153]: * :no_entry_sign: Expected 1 spaces after opening parenthesis; 0 found
* :no_entry_sign: Expected 1 spaces before closing parenthesis; 0 found
* :no_entry_sign: There must be a single space after the first parenthesis
* :no_entry_sign: There must be a single space before the last parenthesis

<sub>hash: 320fd98d5213d70ba308d5b6585f1b11</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 14]
```diff
@@ -110,35 +111,53 @@ public function render_update_message() {
...
+            $this->get_ajax_nonce_action(),
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:154]: :no_entry_sign: Tabs must be used to indent lines; spaces are not allowed

<sub>hash: 702c951d9197fa82e1904392e5a0fd5b</sub>

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 15]
```diff
@@ -110,35 +111,53 @@ public function render_update_message() {
...
+            $this->get_unable_to_continue_processing_data()
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:155]: :no_entry_sign: Tabs must be used to indent lines; spaces are not allowed

<sub>hash: c6da15ad100613a85dd03cef5a2a6e42</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 6 error(s)

**Review by lucatume — CHANGES_REQUESTED** [review]
This is a neat idea, but the Heartbeat API can be disabled.
I would rely on that, if available, and fall back on the existing system if not available.
Probably scaffold a simple API that our current implementation and the Heartbeat API handler can implement (or share, with a trait), remove knowledge from the JS component of the BE system handling the request by localizing, from the backend, something telling it what endpoint/action to use.
Add tests.

**Comment by mitogh** [comment]
Great point :)

File: src/Tribe/Aggregator/Record/Queue_Realtime.php [hunk 16]
```diff
@@ -35,7 +35,9 @@ public function __construct(
 	) {
 		tribe_notice( 'aggregator-update-msg', array( $this, 'render_update_message' ), 'type=warning&dismiss=0' );
 
+		add_filter( 'heartbeat_received', [ $this, 'receive_heartbeat' ], 10, 2 );
 		add_action( 'wp_ajax_tribe_aggregator_realtime_update', array( $this, 'ajax' ) );
```
Comment (tr1b0t) [inline src/Tribe/Aggregator/Record/Queue_Realtime.php:39]: :no_entry_sign: Short array syntax must be used to define arrays

<sub>hash: 7d234b2cc427e8062872a5976d31178c</sub>

**Review by tr1b0t — COMMENTED** [review]
:no_entry_sign: 1 error(s)

File: src/resources/js/aggregator-fields.js [hunk 17]
```diff
@@ -1075,15 +1075,38 @@ tribe_aggregator.fields = {
 		obj.progress.$.bar       = obj.progress.$.notice.find( '.bar' );
 		obj.progress.data.time   = Date.now();
 
+		obj.progress.hasHeartBeat = typeof wp !== 'undefined' && wp.heartbeat;
```
Comment (lucatume) [inline src/resources/js/aggregator-fields.js:1078]: This should probably be a Yoda conditoin.

**Review by lucatume — APPROVED** [review]
Solve conflicts, then good to merge.

**Pull request merged**